### ISSUE TITLE
[11.0] [?10.0?] base_rest: RFC: Annotation for the methods exposed by the services

### ISSUE THREAD

**Issue by lmignon** [OCA/rest-framework#6]
## Current context

The current implementation is based on implicit rules:
* public methods defined into a component inheriting of `'base.rest.service'` are implicitly available into the REST API. 
```python
class PartnerService(Component):
    _inherit = 'base.rest.service'
    _name = 'partner.service'
    _usage = 'partner'
    _collection = 'base.rest.demo.private.services'
    _description = """
Partner Services
    """

    def get(self, _id):
        """
        Get partner's informations
        """
        return self._to_json(self._get(_id))
```
* schemas used to validate the data transferred by these services (in/out) are provided by protected methods (_validator_...). The link between services methods and methods providing validation schemas is based on naming conventions.

```python
    def get(self, _id):
        """
        Get partner's informations
        """
        return self._to_json(self._get(_id))

    # Validator
    def _validator_get(self):
        return {}

    def _validator_return_get(self):
        res = self._validator_create()
        res.update({
            'id': {'type': 'integer', 'required': True, 'empty': False},
        })
        return res
```

* It's possible to return a result in a different format than JSON. Nevertheless there is no way to document/enforce this type of result

```python
from odoo.http import Response
...
    def get_image(self, _id):
        ....
        return Response(
            output.getvalue(),
            headers={
                'Content-Type': mimetype,
                'Content-Disposition': 'inline; filename="%s"' % filename,
            })
```

* By default, the base_rest addon register generic http routes. To avoid the call of methods that could do updates by an HTTP GET request (which is against the HTTP) we enforce only 2 methods available by default via HTTP GET. If you need to add another method, you must declare it into the controller.

* It's not possible to specify that a method takes as input our returns as output a list of items... 

## Proposal

I I think that the development of a specific decorator to annotate the exposed methods would allow us to pass from an implicit mode to an explicit mode for the various stated elements. This would also allow us to have a stronger semantic in the definition of methods.

```python

    @restapi.method(
        input = None,
        output = File(),
        auth='User',
        method='GET'
    )
    def get_image(self, _id):
        ....
        return {
            'mimetype': ...
            'filename': ....
            'data': output.getvalue(),
        }

    @restapi.method(
        input = JSon(schema_method=`_validator_get`, is_list=False, # default ),
        output = JSon(schema_method=`_validator_return_get`,  is_list=True),
        auth='User'
    )
    def get(self, _id):
        """
        Get partner's informations
        """
        return self._get(_id)

    # Validator
    def _validator_get(self):
        return {}

    def _validator_return_get(self):
        res = self._validator_create()
        res.update({
            'id': {'type': 'integer', 'required': True, 'empty': False},
        })
        return res
```

Serialize / deserialize could be done by the decorator....

**Comment by lmignon** [comment]
@sebastienbeau @sbidoul you comments are welcome....

**Comment by dreispt** [comment]
Why not simply:
```
    @restapi.method(
        input=_validator_get,
        output=_validator_return_get,
        auth='user'
    )
    def get(self, _id):
```

**Comment by lmignon** [comment]
@dreispt see https://github.com/OCA/rest-framework/pull/48/files#diff-446e66ad580b33ff05f11e44b939cd38R17
```python
    @restapi.method(
        [(["/<int:id>/get", "/<int:id>"], "GET")],
        output_param=restapi.Datamodel("partner.info"),
        auth="public",
    )
```
The link is for a service using a "DataModel" as param.

For a service using cerberus the code will be:

```python
@restapi.method(
        [(["/<int:id>/get", "/<int:id>"], "GET")],
        output_param=restapi.CerberusValidator(schema="_validator_return_get"),
        auth="public",
    )
``` 
The new API is ready for review.... The old api is still supported.

**Comment by lmignon** [comment]
implemented  in #93 #48 and #71 (all merged)

**Comment by PSLuan** [comment]
When using @restapi.method i'm facing this error:
Function declared as capable of handling request of type 'http' but called with a request of type 'json'

Does any of you would know how to solve it? 
Here is an example of how is my endpoint actually.
@restapi.method([(["/login"], "POST")], auth="public", type="json")
I'm in fact sending a json body on request, but I cant understand why this is happening.
